File: packages/archive/claw-server-ts/src/api/http/handlers/recordings.ts
import type { AppendRecordingEventsResponse } from '@browseros/claw-api'
import type { Handler } from 'hono'
import { getTabTargetMap } from '../../../lib/tab-targets'
import {
  type RecordingEventInput,
  recordingStore,
} from '../../../services/recordings'
import { apiError } from '../errors'
import type { RequestContextEnv } from '../request-context'
import { positiveInteger } from '../validation'

export interface RecordingIdentity {
  tabId: number
  documentId: string
}

export interface RecordingHandlerDependencies {
  appendRecordingEvents(
    identity: RecordingIdentity,
    events: RecordingEventInput[],
    batchId: string,
    hasGap: boolean,
  ): Promise<AppendRecordingEventsResponse>
}

export interface RecordingHandlers {
  ingest: Handler<RequestContextEnv>
}

export function createRecordingHandlers(
  dependencies: RecordingHandlerDependencies,
): RecordingHandlers {
  return {
    ingest: async (c) => {
      const contentType = c.req.header('content-type') ?? ''
      if (!contentType.toLowerCase().startsWith('application/x-ndjson')) {
        return apiError(
          c,
          400,
          'invalid_request',
          'content-type must be application/x-ndjson',
        )
      }
      const tabId = positiveInteger(c.req.header('x-recording-tab-id') ?? '')
      const documentId = c.req.header('x-recording-document-id') ?? ''
      const batchId = c.req.header('x-recording-batch-id') ?? ''
      const hasGap = recordingGap(c.req.header('x-recording-has-gap'))
      if (
        tabId === null ||
        !isChromeDocumentId(documentId) ||
        batchId.length === 0 ||
        hasGap === null
      ) {
        return apiError(
          c,
          400,
          'invalid_request',
          'recording tab, document, batch, and gap headers are invalid',
        )
      }
      const parsed = parseRecordingEvents(await c.req.text())
      if (parsed.events.length === 0) return c.json({ accepted: 0 })
      return c.json(
        await dependencies.appendRecordingEvents(
          { tabId, documentId },
          parsed.events,
          batchId,
          hasGap || parsed.droppedLines > 0,
        ),
      )
    },
  }
}

export const productionRecordingHandlers = createRecordingHandlers({
  async appendRecordingEvents(identity, events, batchId, hasGap) {
    const targetId =
      (await getTabTargetMap()?.targetForTab(identity.tabId)) ?? null
    const appended = await recordingStore.appendBatch({
      documentId: identity.documentId,
      tabId: identity.tabId,
      targetId,
      events,
      batchId,
      hasGap,
    })
    return { accepted: appended ? events.length : 0 }
  },
})

function recordingGap(raw: string | undefined): boolean | null {
  if (raw === undefined || raw === 'false') return false
  if (raw === 'true') return true
  return null
}

function isChromeDocumentId(value: string): boolean {
  return /^[0-9a-f]{32}$/i.test(value)
}

/** Recorder NDJSON is tolerant: malformed lines and events without a finite timestamp are dropped. */
function parseRecordingEvents(ndjson: string): {
  events: RecordingEventInput[]
  droppedLines: number
} {
  const events: RecordingEventInput[] = []
  let droppedLines = 0
  for (const line of ndjson.split('\n')) {
    if (!line.trim()) continue
    try {
      const event = JSON.parse(line) as Record<string, unknown>
      if (typeof event.ts !== 'number' || !Number.isFinite(event.ts)) {
        droppedLines++
        continue
      }
      events.push({ ts: event.ts, type: event.type, data: event.data })
    } catch {
      droppedLines++
    }
  }
  return { events, droppedLines }
}

File: packages/browseros-agent/apps/claw-server-rust/src/api/http/recordings.rs
use super::{error, internal};
use crate::{
    AppState,
    error::{CanonicalError, RequestId},
    services::recordings::RecordingEventInput,
};
use axum::{
    Extension, Json,
    extract::{State, rejection::StringRejection},
    http::{HeaderMap, StatusCode, header},
};
use claw_api::models::AppendRecordingEventsResponse;

pub(super) async fn append_document_events(
    Extension(request_id): Extension<RequestId>,
    State(state): State<AppState>,
    headers: HeaderMap,
    body: Result<String, StringRejection>,
) -> Result<Json<AppendRecordingEventsResponse>, CanonicalError> {
    let body = recording_body(&request_id, body)?;
    require_ndjson(&request_id, &headers)?;
    let tab_id = positive_recording_header(&request_id, &headers, "x-recording-tab-id")?;
    let document_id = required_header(&request_id, &headers, "x-recording-document-id")?;
    let batch_id = required_header(&request_id, &headers, "x-recording-batch-id")?;
    let gap_header = gap_header(&request_id, &headers)?;
    if !is_chrome_document_id(&document_id) {
        return Err(error(
            &request_id,
            StatusCode::BAD_REQUEST,
            "invalid_request",
            "recording tab, document, batch, and gap headers are invalid",
        ));
    }
    let parsed = parse_recording_events(&body);
    let appended = state
        .recording_ingest
        .append_document(
            &document_id,
            tab_id,
            &parsed.events,
            &batch_id,
            gap_header || parsed.dropped_lines > 0,
        )
        .await
        .map_err(|source| internal(&request_id, source))?;
    Ok(Json(AppendRecordingEventsResponse::new(if appended {
        i64::try_from(parsed.events.len()).unwrap_or(i64::MAX)
    } else {
        0
    })))
}

/// Tolerant parse of recorder-supplied NDJSON: lines that are not JSON
/// or lack an integer `ts` are dropped, never fatal.
struct ParsedRecordingEvents {
    events: Vec<RecordingEventInput>,
    dropped_lines: usize,
}

fn parse_recording_events(body: &str) -> ParsedRecordingEvents {
    let mut events = Vec::new();
    let mut dropped_lines = 0;
    for line in body.lines().filter(|line| !line.trim().is_empty()) {
        let Ok(event) = serde_json::from_str::<serde_json::Value>(line) else {
            dropped_lines += 1;
            continue;
        };
        let Some(ts) = event.get("ts").and_then(serde_json::Value::as_i64) else {
            dropped_lines += 1;
            continue;
        };
        events.push(RecordingEventInput {
            ts,
            event_type: event.get("type").cloned(),
            data: event.get("data").cloned(),
        });
    }
    ParsedRecordingEvents {
        events,
        dropped_lines,
    }
}

fn recording_body(
    request_id: &RequestId,
    body: Result<String, StringRejection>,
) -> Result<String, CanonicalError> {
    body.map_err(|rejection| {
        if rejection.status() == StatusCode::PAYLOAD_TOO_LARGE {
            error(
                request_id,
                StatusCode::PAYLOAD_TOO_LARGE,
                "recording_payload_too_large",
                &format!(
                    "recording payload exceeds {} byte limit",
                    super::RECORDING_INGEST_MAX_BYTES
                ),
            )
        } else {
            error(
                request_id,
                StatusCode::BAD_REQUEST,
                "invalid_request",
                "recording payload must be valid UTF-8",
            )
        }
    })
}

fn require_ndjson(request_id: &RequestId, headers: &HeaderMap) -> Result<(), CanonicalError> {
    let valid = headers
        .get(header::CONTENT_TYPE)
        .and_then(|value| value.to_str().ok())
        .is_some_and(|value| {
            value
                .to_ascii_lowercase()
                .starts_with("application/x-ndjson")
        });
    if valid {
        Ok(())
    } else {
        Err(error(
            request_id,
            StatusCode::BAD_REQUEST,
            "invalid_request",
            "content-type must be application/x-ndjson",
        ))
    }
}

fn required_header(
    request_id: &RequestId,
    headers: &HeaderMap,
    name: &str,
) -> Result<String, CanonicalError> {
    headers
        .get(name)
        .and_then(|value| value.to_str().ok())
        .filter(|value| !value.is_empty())
        .map(str::to_string)
        .ok_or_else(|| {
            error(
                request_id,
                StatusCode::BAD_REQUEST,
                "invalid_request",
                "recording tab, document, batch, and gap headers are invalid",
            )
        })
}

fn gap_header(request_id: &RequestId, headers: &HeaderMap) -> Result<bool, CanonicalError> {
    match headers
        .get("x-recording-has-gap")
        .and_then(|value| value.to_str().ok())
    {
        None | Some("false") => Ok(false),
        Some("true") => Ok(true),
        Some(_) => Err(error(
            request_id,
            StatusCode::BAD_REQUEST,
            "invalid_request",
            "recording tab, document, batch, and gap headers are invalid",
        )),
    }
}

fn is_chrome_document_id(value: &str) -> bool {
    value.len() == 32 && value.bytes().all(|byte| byte.is_ascii_hexdigit())
}

fn positive_recording_header(
    request_id: &RequestId,
    headers: &HeaderMap,
    name: &str,
) -> Result<i64, CanonicalError> {
    let value = headers
        .get(name)
        .and_then(|value| value.to_str().ok())
        .and_then(|value| value.parse::<i64>().ok())
        .filter(|value| *value > 0)
        .ok_or_else(|| {
            error(
                request_id,
                StatusCode::BAD_REQUEST,
                "invalid_request",
                "recording tab, page, and target headers are required",
            )
        })?;
    Ok(value)
}
